strawberryfields/configuration.py

```diff
@@ -15,28 +15,136 @@
 }
+"""dict: Nested dictionary representing the allowed configuration
+sections, options, default values, and allowed types for Strawberry
+Fields configurations. For each configuration option key, the
+corresponding value is a length-2 tuple, containing:
+
+* A type or tuple of types, representing the allowed type
+  for that configuration option.
+
+* The default value for that configuration option.
+
+.. note::
+
+    By TOML convention, keys with a default value of ``None``
+    will **not** be present in the generated/loaded configuration
+    file. This is because TOML has no concept of ``NoneType`` or ``Null``,
+    instead, the non-presence of a key indicates that the configuration
+    value is not set.
+"""
 
 
 class ConfigurationError(Exception):
     """Exception used for configuration errors"""
 
 
-def load_config(filename="config.toml", **kwargs):
+def _deep_update(source, overrides):
+    """Recursively update a nested dictionary.
+
+    This function is a generalization of Python's built in
+    ``dict.update`` method, modified to recursively update
+    keys with nested dictionaries.
+    """
+    for key, value in overrides.items():
+        if isinstance(value, collections.Mapping) and value:
+            # Override value is a non-empty dictionary.
+            # Update the source key with the override dictionary.
+            returned = _deep_update(source.get(key, {}), value)
+            source[key] = returned
+        elif value != {}:
+            # Override value is not an empty dictionary.
+            source[key] = overrides[key]
+    return source
+
+
+def _generate_config(config_spec, **kwargs):
+    """Generates a configuration, given a Strawberry Fields configuration
+    specification.
+
+    See :attr:`~.DEFAULT_CONFIG_SPEC` for an example of a valid configuration
+    specification.
+
+    Optional keyword arguments may be provided to override default values
+    in the cofiguration specification. If the provided override values
+    do not match the expected type defined in the configuration spec,
+    a ``ConfigurationError`` is raised.
+
+    **Example**
+
+    >>> _generate_config(DEFAULT_CONFIG_SPEC, api={"port": 54})
+    {
+        "api": {
+            "authentication_token": "",
+            "hostname": "platform.strawberryfields.ai",
+            "use_ssl": True,
+            "port": 54,
+        }
+    }
+
+    Args:
+        config_spec (dict): nested dictionary representing the
+            configuration specification
+
+    Keyword Args:
+        Provided keyword arguments may overwrite default values of
+        matching keys.
+
+    Returns:
+        dict: the default configuration defined by the input config spec
+
+    Raises:
+        ConfigurationError: if provided keyword argument overrides do not
+        match the expected type defined in the configuration spec.
+    """
+    res = {}
+    for k, v in config_spec.items():
```

Comment:
Is there any use case for being able to pass `config_spec` to this function? Most of the times one would be using `DEFAULT_CONFIG_SPEC`, but now the function allows any nested dictionary structure. Not sure if the complexity arising from this is worth having.

To me, it seems that the block within the `isinstance(v, tuple)` case could be renamed and organized into a new function. Since we know that the maximum depth of nesting is 2, using recursion here seems like a general solution to a specific problem which can still be avoided to yield a simpler line of logic.

Reply:
Ah, good point 🙂 I tried to keep this as general as possible, because TOML allows for nested sections. For example, in the future, we might have something like this:

```toml
[api]
hostname = "platform.strawberryfields.ai"

  [default]
  token= "personal_token"

  [org1]
  token= "organization_token"
```

When loaded via `toml.load`, it would look like this:

```python
{
  "api": {
    "hostname": "platform.strawberryfields.ai",
    "default": {"token": "personal_token"},
    "org1": {"token": "organization_token"}
  }
}
```

I suppose it's a question of do we want to disallow nested TOML config files (in which case it makes sense to restrict to simply using `DEFAULT_CONFIG_SPEC` here), or if we want to keep this function general.

Reply:
That's a good point, yes :) I'd be leaning towards keeping it simple for now and updating it whenever a specific nested section is introduced for the configuration (but this is subjective point of view). Overall, this would come with the

- Advantage: the configuration related code is tailored onto the current needs, is slightly simpler and easier to be picked up
- Disadvantage: once a nested section is being added, the logic here will have to change to this more general solution